docs/source/interactive/autoawait.rst

```diff
@@ -0,0 +1,210 @@
+.. _autoawait:
+
+Asynchronous in REPL: Autoawait
+===============================
+
+Starting with IPython 7.0, and when user Python 3.6 and above, IPython offer the
+ability to run asynchronous code from the REPL. Constructs which are
+:exc:`SyntaxError` s in the Python REPL can be used seamlessly in IPython.
+
+The example given here are for terminal IPython, running async code in a
+notebook interface or any other frontend using the Jupyter protocol will need to
+use a newer version of IPykernel. The details of how async code runs in
+IPykernel will differ between IPython, IPykernel and their versions.
+
+When a supported library is used, IPython will automatically `await` Futures and
+Coroutines in the REPL. This will happen if an :ref:`await <await>` (or any
+other async constructs like async-with, async-for) is use at top level scope, or
+if any structure valid only in `async def
+<https://docs.python.org/3/reference/compound_stmts.html#async-def>`_ function
+context are present. For example, the following being a syntax error in the
+Python REPL::
+
+    Python 3.6.0 
+    [GCC 4.2.1]
+    Type "help", "copyright", "credits" or "license" for more information.
+    >>> import aiohttp
+    >>> result = aiohttp.get('https://api.github.com')
+    >>> response = await result
+      File "<stdin>", line 1
+        response = await result
+                              ^
+    SyntaxError: invalid syntax
+
+Should behave as expected in the IPython REPL::
+
+    Python 3.6.0
+    Type 'copyright', 'credits' or 'license' for more information
+    IPython 7.0.0 -- An enhanced Interactive Python. Type '?' for help.
+
+    In [1]: import aiohttp
+       ...: result = aiohttp.get('https://api.github.com')
+
+    In [2]: response = await result
+    <pause for a few 100s ms>
+
+    In [3]: await response.json()
+    Out[3]:
+    {'authorizations_url': 'https://api.github.com/authorizations',
+     'code_search_url': 'https://api.github.com/search/code?q={query}...',
+    ...
+    }
+
+
+You can use the ``c.InteractiveShell.autoawait`` configuration option and set it
+to :any:`False` to deactivate automatic wrapping of asynchronous code. You can also
+use the :magic:`%autoawait` magic to toggle the behavior at runtime::
+
+    In [1]: %autoawait False
+
+    In [2]: %autoawait
+    IPython autoawait is `Off`, and set to use `IPython.core.interactiveshell._asyncio_runner`
```

Comment:
This seems like it is exposing an implementation detail. it would probably be worth defining a `__str__` or something on the runners so that they reflect more closely how they were activated.